src/socket/udp_socket.cr

```diff
@@ -48,113 +51,35 @@ require "./ip_socket"
 # end
 # ```
 class UDPSocket < IPSocket
-  def initialize(@family : Family = Family::INET)
-    super create_socket(family.value, Type::DGRAM, Protocol::UDP)
+  def initialize(family : Family = Family::INET)
+    super(family, Type::DGRAM, Protocol::UDP)
   end
 
-  # Binds the UDP socket to a local address.
+  # Receives a text message from the previously bound address.
   #
   # ```
   # server = UDPSocket.new
-  # server.bind "localhost", 1234
-  # ```
-  def bind(host, port, dns_timeout = nil)
-    getaddrinfo(host, port, @family, Type::DGRAM, Protocol::UDP, timeout: dns_timeout) do |addrinfo|
-      self.reuse_address = true
-
-      ret =
-        {% if flag?(:freebsd) || flag?(:openbsd) %}
-          LibC.bind(fd, addrinfo.ai_addr.as(LibC::Sockaddr*), addrinfo.ai_addrlen)
-        {% else %}
-          LibC.bind(fd, addrinfo.ai_addr, addrinfo.ai_addrlen)
-        {% end %}
-      unless ret == 0
-        next false if addrinfo.ai_next
-        raise Errno.new("Error binding UDP socket at #{host}:#{port}")
-      end
-
-      true
-    end
-  end
-
-  # Connects the UDP socket to a remote address to send messages to.
+  # server.bind("localhost", 1234)
   #
+  # message, client_addr = server.receive
   # ```
-  # client = UDPSocket.new
-  # client.connect("localhost", 1234)
-  # client.send("a text message")
-  # ```
-  def connect(host, port, dns_timeout = nil, connect_timeout = nil)
-    getaddrinfo(host, port, @family, Type::DGRAM, Protocol::UDP, timeout: dns_timeout) do |addrinfo|
-      if err = nonblocking_connect host, port, addrinfo, timeout: connect_timeout
-        next false if addrinfo.ai_next
-        raise err
-      end
-
-      true
-    end
+  def receive(max_message_size = 512) : {String, IPAddress}
+    bytes = Bytes.new(max_message_size)
+    bytes_read, sockaddr, addrlen = recvfrom(bytes)
+    {String.new(bytes.to_unsafe, bytes_read), IPAddress.from(sockaddr, addrlen)}
```

Comment:
Here there's a double allocation: `Bytes.new` and `String.new` (it copies the bytes). You can do something like:

```cr
str = String.new(max_message_size) do |buffer|
  bytes = Slice.new(buffer, max_message_size)
  bytes_read, sockaddr, addrlen = recvfrom(bytes)
  # etc.
  {bytes_read, 0}
end
```

The only "problem" with this is that this will maybe allocate more bytes than needed, so maybe occupy more memory until the string is GC'ed. However, I'll soon change `String#new` to `realloc` the pointer if the resulting bytesize is less than the requested capacity. I even did a few benchmarks and this makes a program go much faster as later less memory needs to be reclaimed.